### DISCUSSION TITLE
Include "score" in the list of anonymized entities returned by anonymize

### DISCUSSION THREAD

**Discussion by dbush-bdq** [microsoft/presidio#680] [Requested features]
Would it be possible to include the "score" in the list of anonymized entities returned by the call to anonymize? 

Also, if provided with the analysis_explaination in the call, flow that through information into the returned items too? 

**Update:**
Just thought of something else, the start and end positions in the original text would also be useful.

**Comment by omri374** [comment]
Hi @dbush-bdq, could you please provide some more context on your scenario?
Deidentification in Presidio is a two-phase process. First, the presidio-analyzer returns the decision process (analysis_explanation) and the indices for detected PII in the original text. Then, the presidio-anonymizer returns the anonymized text. The responses of the two services will provide you with the points you mentioned, so further clarification is needed on the specific case.

**Reply by dbush-bdq** [reply]
Hi @omri374, 

Thanks for the reply. Sorry, on re-reading my post it isn't very clear what I mean. 

I understand that the process is that the results generated by presidio-analyzer are passed along with the original text to presidio-anonymizer to generate the anonymized text, but in addition to this text, presidio-analyzer also returns the array of anonymized items, i.e. operator, entity_type, start, end and text. 

From the API documentation (swapped order of the items for clarity):

```
{
	"text": "hello world, my name is ANONYMIZED. My number is: ANONYMIZED",
	"items": [
		{
			"operator": "replace",
			"entity_type": "NAME",
			"start": 24,
			"end": 34,
			"text": "ANONYMIZED"
		},
		{
			"operator": "replace",
			"entity_type": "PHONE_NUMBER",
			"start": 50,
			"end": 60,
			"text": "ANONYMIZED"
		}
	]
}
```

As part of the anonymization process, the results are consolidated as described in [Overlapping Anonymization Scenarios](https://microsoft.github.io/presidio/anonymizer/#overlapping-anonymization-scenarios). 

For my use, this consolidated set is actually more useful than the overall set returned by presidio-analyzer. It contains fewer results to review and they are more significant, e.g. an email address is more important than the domain name that is contained within it. 

I would like to be able to show alongside these consolidated results the score from presidio-anonymizer because this allows the user to understand how certain a specific result is. 

I could emulate the consolidation logic in my code and stitch this back together but if possible it would be great to extend the items returned by the call to presidio-anonymizer to pass the score through, something like this:

```
{
	"text": "hello world, my name is ANONYMIZED. My number is: ANONYMIZED",
	"items": [
		{
			"operator": "replace",
			"entity_type": "NAME",
			"start": 24,
			"end": 34,
			"text": "ANONYMIZED",
			"score": 0.85
		},
		{
			"operator": "replace",
			"entity_type": "PHONE_NUMBER",
			"start": 50,
			"end": 60,
			"text": "ANONYMIZED",
			"score": 0.6499999999999999
		}
	]
}
```

Further, if possible would be great to also pass through the analysis_explaination for the same reasons. 

Also, the returned item contains the start and end positions for where the item is located in the anonymized text. 

This is useful but in addition, it would be good to know what the start and end positions are in the original text. This is the location where the user may have to go and amend the text to get rid of the PII issue. If the text is long, identifying where this is can be difficult. 

So, if the input string was "hello world, my name is Dominic Bush. My number is: 00 44 123456789" the output would look something like this:

```
{
	"text": "hello world, my name is ANONYMIZED. My number is: ANONYMIZED",
	"items": [
		{
			"operator": "replace",
			"entity_type": "NAME",
			"start": 24,
			"end": 34,
			"text": "ANONYMIZED",
			"score": 0.85,
			"analysis_explanation": { ... lots of stuff ...},
			"source_start": 24,
			"source_end": 36
		},
		{
			"operator": "replace",
			"entity_type": "PHONE_NUMBER",
			"start": 50,
			"end": 60,
			"text": "ANONYMIZED",
			"score": 0.6499999999999999,
			"analysis_explanation": { ... lots of stuff ...}
			"source_start": 52,
			"source_end": 67
		}
	]
}
```

Thanks again for taking a look at this.

**Reply by navalev** [reply]
Hi @dbush-bdq, presidio-anonymizer does not return the "source" position, the decision what entity to anonymize is explained here https://microsoft.github.io/presidio/anonymizer/#overlapping-anonymization-scenarios. 
We welcome community contributions, and would be happy to support you if you want to contribute to presidio!

**Reply by dbush-bdq** [reply]
Hi @navalev, 

Thanks for the reply. 

I was suggesting adding the "source" positions as a feature request and I also referenced the link you mention in my earlier post.

Understood regarding community contributions. I'll see if I can contribute.

Thanks,
Dom.
